### DISCUSSION TITLE
WeightedEdge in case of Ecore model

### DISCUSSION THREAD

**Discussion by hsd-dev** [eclipse-glsp/glsp#684] [Q&A]
If my source model is Ecore based, do I need to update the notation model as well similar to the graph model to add properties to the edges? I want to add properties (a struct) to my edge. Once I make a connection between source and target, how do I edit these properties?

Also, in the workflow example, how do I edit the edge properties to change the probability?

**Comment by hsd-dev** [comment]
I have updated the `java-emf-theia` example to add a weighted transition. What I have so far:
### server-side
`tasks-graph.ecore`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ecore:EPackage xmi:version="2.0" xmlns:xmi="http://www.omg.org/XMI" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns:ecore="http://www.eclipse.org/emf/2002/Ecore" name="tasklistgraph" nsURI="http://www.eclipse.org/glsp/examples/tasklist/graph"
    nsPrefix="tasklistgraph">
  <eClassifiers xsi:type="ecore:EClass" name="WeightedTransition" eSuperTypes="glsp-graph.ecore#//GEdge">
    <eStructuralFeatures xsi:type="ecore:EAttribute" name="weight" lowerBound="1"
        eType="ecore:EDataType http://www.eclipse.org/emf/2002/Ecore#//EDouble"/>
  </eClassifiers>
</ecore:EPackage>
```
`TransitionEdgeBuilder.java`
```java
public static class TransitionEdgeBuilder extends AbstractGEdgeBuilder<WeightedTransition, TransitionEdgeBuilder> {
     ...
      public TransitionEdgeBuilder() {
         super(TaskListModelTypes.WEIGHTED_TRANSITION);
         // defined in TaskListModelTypes
         // public static final String WEIGHTED_TRANSITION = "edge:weighted";
      }
...
```

updated `TaskListGModelFactory.java`
```java
protected GEdge createTransitionEdge(final Transition transition, final GGraph graph) {
      String sourceId = transition.getSource().getId();
      String targetId = transition.getTarget().getId();

      GModelElement sourceNode = findGNodeById(graph.getChildren(), sourceId);
      GModelElement targetNode = findGNodeById(graph.getChildren(), targetId);

      TransitionEdgeBuilder transitionEdgeBuilder = new TransitionEdgeBuilder().source(sourceNode)
         .target(targetNode)
         .setWeight(0.5)
         .id(idGenerator.getOrCreateId(transition));
      applyEdgeData(transition, transitionEdgeBuilder);
      return transitionEdgeBuilder.build();
   }

   private TransitionEdgeBuilder applyEdgeData(final EObject edgeElement, final TransitionEdgeBuilder builder) {
      // TODO Auto-generated method stub
      modelState.getIndex().getNotation(edgeElement, Edge.class)
         .ifPresent(edge -> applyEdgeData(edge, builder));
      return builder;
   }

   private static TransitionEdgeBuilder applyEdgeData(final Edge edge, final TransitionEdgeBuilder builder) {
      if (edge.getBendPoints() != null) {
         edge.getBendPoints().stream().map(GraphUtil::copy).forEachOrdered(builder::addRoutingPoint);
      }
      return builder;
   }
```
Registered `TaskListGraphExtension.java` in `TaskListDiagramModule`

### client-side
```ts
const tasklistDiagramModule = new ContainerModule((bind, unbind, isBound, rebind) => {
    ...
    configureModelElement(context, 'edge:weighted', WeightedTransition, WeightedTransitionView);
});
```
Am I on the right path? What else needs to be updated?

**Reply by hsd-dev** [reply]
OK I edited the source model as well by adding a weight attribute to `Transition` and set a default weight in `createTransition` in `CreateTransitionEdgeHandler`. Now I am able to create transitions with default weights. But coming back to my original question: how do I edit the weights of these transitions / edges now? I am guessing I need to add a "edit edge handler"?

**Comment by martin-fleck-at** [comment]
Hi @ipa-hsd,

In order to edit properties you will need a few things working together:
- client: How you want to render the value that you want to edit on the client side (GModel representation and rendering)
- protocol: A GLSP operation that represents your edit data that will be sent from the client to the server
- server: a registered operation handler on the GLSP server side that updates the internal data source (EMF model in our case) and sends back an updated GModel to the client

In the workflow example, we use two different ways to approach editing: You can edit the name directly on the label by double clicking it (using the generic label edit functionality) and you can edit properties of a task by right clicking a task and selecting "Direct Edit Task" ([TaskEditor](https://github.com/eclipse-glsp/glsp-examples/blob/7790d1abfe4ee79ec9f7784adb59b125e2cfd0e8/workflow/glsp-client/workflow-glsp/src/direct-task-editing/direct-task-editor.ts) on the client, [ApplyTaskEditOperationHandler](https://github.com/eclipse-glsp/glsp-examples/blob/7790d1abfe4ee79ec9f7784adb59b125e2cfd0e8/workflow/glsp-server/src/main/java/org/eclipse/glsp/example/workflow/taskedit/ApplyTaskEditOperationHandler.java) on the server). Please note that currently there is no way to directly edit the weight of weighted edges but you could follow a similar approach as outlined before. Alternatively, you can also use an "external" UI such as a [properties view](https://github.com/eclipse-theia/theia/tree/master/packages/property-view) or a tree view as done in the [coffee editor example](https://www.eclipse.org/emfcloud/#coffeeeditoroverview) ([online demonstration](https://eclipsesource.com/coffee-editor)). 

On the server side you need a handler that handles the operation that you just sent. For the EMF case, the basic idea is that we do not manipulate the GModel directly (as, for instance, the [generic label editing handler]((https://github.com/eclipse-glsp/glsp-server/blob/b6822d5122dc2700ec2811a4c64f7745df832cb0/plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/gmodel/GModelApplyLabelEditOperationHandler.java)) does) but instead we manipulate our EMF resources (semantic + notation) using EMF commands. After the operation is executed, the EMF resources are translated into a GModel using the GModelFactory where any properties that are stored in the EMF models can be translated into GModel elements or properties. For the task list example, there are two handlers for [creation](https://github.com/eclipse-glsp/glsp-examples/blob/5aecd0dd10db9bb745bb0d3922612a4120961310/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/CreateTaskNodeHandler.java) and [deletion](https://github.com/eclipse-glsp/glsp-examples/blob/5aecd0dd10db9bb745bb0d3922612a4120961310/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/handler/DeleteTaskNodeHandler.java). Regarding the translation from EMF resources to the GModel we generically only apply basic shape data for nodes and translate bend points into routing points for edges.

So for your use case that means that you should extend your EMF model with the weight property (e.g., Transition in the model.ecore in the example) and adapt the translation from the EMF model to the GModel to consider the new property. How you translate the weight really depends on your use case. If you use the label editing mechanism you probably will create a GLabel but if use another approach maybe simply adding an argument to the generic args map of an GEdge may be sufficient. After you configure the respective rendering on the client side you already have a static support for the weight property. For editing you then need to create an operation and operation handler. The handler will use the data to find the correct EMF model element from the index and apply an EMF command on it that adapts the weight. If the GModelFactory was already rendering the weight property correctly before nothing more needs to be done.

**Reply by hsd-dev** [reply]
Thanks for the detailed answer @martin-fleck-at! I think property view is what I was looking for. Even the default view is fine for me as of now. Just for reference for others, leaving the doc for property view here: https://theia-ide.org/docs/property_view/

**Reply by hsd-dev** [reply]
to test `@theia/property-view`, I updated the `minimal` example with
```
const minimalDiagramModule = new ContainerModule((bind, unbind, isBound, rebind) => {
    rebind(TYPES.ILogger).to(ConsoleLogger).inSingletonScope();
    rebind(TYPES.LogLevel).toConstantValue(LogLevel.warn);
    const context = { bind, unbind, isBound, rebind };
    configureDefaultModelElements(context);

    bind(PropertyViewWidgetProvider).to(EmptyPropertyViewWidgetProvider);
});
```
but I don't see the `Property` view in theia.

**Reply by hsd-dev** [reply]
When I include `"@theia/property-view": "^1.0.0",` in `minimal-browser-app/package.json`, I get the error
```
root ERROR Error: It looks like there is a circular dependency in one of the 'toDynamicValue' bindings. Please investigate bindings withservice identifier 'class ResourcePropertyViewTreeWidget extends browser_1.TreeWidget {
    constructor(props, model, contextMenuRenderer) {
        super(props, model, contextMenuRenderer);
        model.root = {
            id: resource_property_view_tree_items_1.ROOT_ID,
            name: ResourcePropertyViewTreeWidget_1.LABEL,
            parent: undefined,
            visible: false,
            children: []
        };
        this.propertiesTree = new Map();
    }
    init() {
        super.init();
        this.id = ResourcePropertyViewTreeWidget_1.ID + '-treeContainer';
        this.addClass('treeContainer');
        this.fillPropertiesTree();
    }
    updateNeeded(selection) {
        return this.currentSelection !== selection;
    }
    updatePropertyViewContent(propertyDataService, selection) {
        if (this.updateNeeded(selection)) {
            this.currentSelection = selection;
            if (propertyDataService) {
                propertyDataService.providePropertyData(selection).then((fileStatObject) => {
                    this.fillPropertiesTree(fileStatObject);
                });
            }
        }
    }
    fillPropertiesTree(fileStatObject) {
        if (fileStatObject) {
            this.propertiesTree.clear();
            const infoNode = this.createCategoryNode('info', nls_1.nls.localizeByDefault('Info'));
            this.propertiesTree.set('info', infoNode);
            infoNode.children.push(this.createResultLineNode('isDirectory', nls_1.nls.localize('theia/property-view/directory', 'Directory'), fileStatObject.isDirectory, infoNode));
            infoNode.children.push(this.createResultLineNode('isFile', nls_1.nls.localizeByDefault('File'), fileStatObject.isFile, infoNode));
            infoNode.children.push(this.createResultLineNode('isSymbolicLink', nls_1.nls.localize('theia/property-view/symbolicLink', 'Symbolic link'), fileStatObject.isSymbolicLink, infoNode));
            infoNode.children.push(this.createResultLineNode('location', nls_1.nls.localize('theia/property-view/location', 'Location'), this.getLocationString(fileStatObject), infoNode));
            infoNode.children.push(this.createResultLineNode('name', nls_1.nls.localizeByDefault('Name'), this.getFileName(fileStatObject), infoNode));
            infoNode.children.push(this.createResultLineNode('path', nls_1.nls.localize('theia/property-view/path', 'Path'), this.getFilePath(fileStatObject), infoNode));
            infoNode.children.push(this.createResultLineNode('lastModification', nls_1.nls.localize('theia/property-view/lastModified', 'Last modified'), this.getLastModificationString(fileStatObject), infoNode));
            infoNode.children.push(this.createResultLineNode('created', nls_1.nls.localize('theia/property-view/created', 'Created'), this.getCreationTimeString(fileStatObject), infoNode));
            infoNode.children.push(this.createResultLineNode('size', nls_1.nls.localize('theia/property-view/size', 'Size'), this.getSizeString(fileStatObject), infoNode));
            this.refreshModelChildren();
        }
    }
    getLocationString(fileStat) {
        return fileStat.resource.path.fsPath();
    }
    getFileName(fileStat) {
        return this.labelProvider.getName(fileStat.resource);
    }
    getFilePath(fileStat) {
        return this.labelProvider.getLongName(fileStat.resource);
    }
    getLastModificationString(fileStat) {
        return fileStat.mtime ? new Date(fileStat.mtime).toLocaleString() : '';
    }
    getCreationTimeString(fileStat) {
        return fileStat.ctime ? new Date(fileStat.ctime).toLocaleString() : '';
    }
    getSizeString(fileStat) {
        return fileStat.size ? nls_1.nls.localizeByDefault('{0}B', fileStat.size.toString()) : '';
    }
    /*
    * Creating TreeNodes
    */
    createCategoryNode(categoryId, name) {
        return {
            id: categoryId,
            parent: this.model.root,
            name,
            children: [],
            categoryId,
            selected: false,
            expanded: true
        };
    }
    createResultLineNode(id, name, property, parent) {
        return {
            id: `${parent.id}::${id}`,
            parent,
            name: name,
            property: property !== undefined ? String(property) : '',
            selected: false
        };
    }
    /**
     * Rendering
     */
    async refreshModelChildren() {
        if (resource_property_view_tree_items_1.ResourcePropertiesRoot.is(this.model.root)) {
            this.model.root.children = Array.from(this.propertiesTree.values());
            this.model.refresh();
        }
    }
    renderCaption(node, props) {
        if (resource_property_view_tree_items_1.ResourcePropertiesCategoryNode.is(node)) {
            return this.renderExpandableNode(node);
        }
        else if (resource_property_view_tree_items_1.ResourcePropertiesItemNode.is(node)) {
            return this.renderItemNode(node);
        }
        return undefined;
    }
    renderExpandableNode(node) {
        return React.createElement(React.Fragment, null,
            React.createElement("div", { className: `theia-resource-tree-node-icon ${this.toNodeIcon(node)}` }),
            React.createElement("div", { className: 'theia-resource-tree-node-name theia-TreeNodeSegment theia-TreeNodeSegmentGrow' }, this.toNodeName(node)));
    }
    renderItemNode(node) {
        return React.createElement(React.Fragment, null,
            React.createElement("div", { className: `theia-resource-tree-node-icon ${this.toNodeIcon(node)}` }),
            React.createElement("div", { className: 'theia-resource-tree-node-name theia-TreeNodeSegment theia-TreeNodeSegmentGrow' }, this.toNodeName(node)),
            React.createElement("div", { className: 'theia-resource-tree-node-property theia-TreeNodeSegment theia-TreeNodeSegmentGrow' }, this.toNodeDescription(node)));
    }
    createNodeAttributes(node, props) {
        return Object.assign(Object.assign({}, super.createNodeAttributes(node, props)), { title: this.getNodeTooltip(node) });
    }
    getNodeTooltip(node) {
        if (resource_property_view_tree_items_1.ResourcePropertiesCategoryNode.is(node)) {
            return this.labelProvider.getName(node);
        }
        else if (resource_property_view_tree_items_1.ResourcePropertiesItemNode.is(node)) {
            return `${this.labelProvider.getName(node)}: ${this.labelProvider.getLongName(node)}`;
        }
        return undefined;
    }
}'.
    at invokeFactory (http://localhost:3000/bundle.js:79181:19)
    at http://localhost:3000/bundle.js:79232:26
    at Object.resolve (http://localhost:3000/bundle.js:79265:12)
    at http://localhost:3000/bundle.js:78284:37
    at Container.__webpack_modules__.../node_modules/inversify/lib/container/container.js.Container._get (http://localhost:3000/bundle.js:78275:44)
    at Container.__webpack_modules__.../node_modules/inversify/lib/container/container.js.Container.get (http://localhost:3000/bundle.js:78192:21)
    at createResourcePropertyViewTreeWidget (http://localhost:3000/vendors-node_modules_theia_property-view_lib_browser_property-view-frontend-module_js.bundle.js:978:18)
    at Binding.dynamicValue (http://localhost:3000/vendors-node_modules_theia_property-view_lib_browser_property-view-frontend-module_js.bundle.js:984:101)
    at http://localhost:3000/bundle.js:79232:118
    at invokeFactory (http://localhost:3000/bundle.js:79177:16)
```
The example [person-detail-property-view](https://github.com/eclipse-emfcloud/jsonforms-property-view/tree/master/example/person-detail-property-view) has worked for me without problems, so maybe there is a conflict somewhere with `eclipse-glsp`?

**Comment by martin-fleck-at** [comment]
@ipa-hsd I can reproduce the problem and it seems to be a version conflict indeed. The current minimal-example uses Theia 1.20.0 but if you specify the dependency of "^1.0.0" you allow much newer versions of that package as well (+ all their dependencies), so there might be some Theia dependencies that are included several times in different versions (once in 1.20.0 and once in the latest version) causing the mismatch. You might even see warnings like:
```
warning Resolution field "@theia/core@1.20.0" is incompatible with requested version "@theia/core@1.26.0"
```

If you specify the matching version of "1.20.0" for the property-view, I believe it should work.

**Reply by hsd-dev** [reply]
great! That works.
